csrank/objectranking/lambdamart.py

```diff
@@ -0,0 +1,370 @@
+import logging
+from collections import deque
+from multiprocessing import Pool
+
+import numpy as np
+from sklearn.tree import DecisionTreeRegressor
+
+from csrank.learner import Learner
+from csrank.metrics import ndcg_at_k
+from csrank.objectranking.object_ranker import ObjectRanker
+
+
+class LambdaMART(ObjectRanker,Learner):
+    def __init__(self, n_objects=None, n_object_features=None, number_of_trees=5, learning_rate=0.1,
+                 min_samples_split=2, max_depth=50, min_samples_leaf=1, max_leaf_nodes=None):
+        """
+            Create a LambdaMART based rank regression model. This model uses an ensemble of trees that learn to predict
+            the relevance scores of the documents based on the features, which then can be turned into rankings.
+            The base learner used is the implementation of Decision Tree from the sklearn tree package. The learner
+            tries to indirectly optimize the nDCG metric by learning the lambdas.
+
+            Parameters
+            ----------
+            n_object_features : int
+                Number of features of the object space
+            n_objects : int
+                Number of objects
+            number_of_trees : int
+                The maximum number of trees that are to be trained for the ensemble.
+            learning_rate : float
+                learning rate for the LambdaMART algorithm
+            min_samples_split : int
+                Number of samples required to split the internal node
+            max_depth : int
+                Maximum depth of the tree
+            min_samples_leaf : int
+                Minimum number of samples required to be at the leaf node
+        """
+        self.n_object_features = n_object_features
+        self.n_objects = n_objects
+        self.number_of_trees = number_of_trees
+        self.learning_rate = learning_rate
+        self.min_samples_split = min_samples_split
+        self.max_depth = max_depth
+        self.min_samples_leaf = min_samples_leaf
+        self.max_leaf_nodes = max_leaf_nodes
+        self.trees = []
+        self.logger = logging.getLogger(LambdaMART.__name__)
+
+    def _custom_letor_dataset_reader(self, filepath):
```

Comment:
This should not be part of the learner.
The learner does not know, what datasets it is given.